webassembly/Dockerfile

```diff
@@ -0,0 +1,20 @@
+FROM gitpod/workspace-full:latest
+
+USER gitpod
+
+RUN bash -cl "cargo install wasm-pack cargo-wasm cargo-generate \
+    && curl \"https://wasmtime.dev/install.sh\" -sSf | bash"
+
+RUN wget "https://github.com/CraneStation/wasi-sdk/releases/download/wasi-sdk-8/wasi-sdk_8.0_amd64.deb" \
+    && sudo dpkg -i wasi-sdk_8.0_amd64.deb
+
+
+RUN bash -cl "git clone \"https://github.com/emscripten-core/emsdk.git\" \
+    && cd emsdk \
+    && ./emsdk install latest \
+    && ./emsdk activate latest \
+    && source ./emsdk_env.sh"
+
+RUN bash -cl "echo -e \"\nsource /home/gitpod/emsdk/emsdk_env.sh\nclear\" >> ~/.bashrc"
+
+ENV PATH="/opt/wasi-sdk/bin:$PATH"
```

Comment:
Hi @bnjbvr! 👋

Quick question: 😇 I have almost no idea what web assembly development looks like. 😬 But knowing that the base image `workspace-full` already has many tools like latest Rust, Clang, GDB, etc, do the tools installed in this Dockerfile seem useful to you, for people intending to work with web assembly? (Using any source language.)

- Are these tools useful for building stuff with wasm?
- Should other useful tools probably be installed as well?

Reply:
Hi there! Great to see helpers for wasm development!

I don't usually develop *in* wasm myself, since I'm mostly focused on developing wasm itself, but this looks like a good building block for C++ development to target WebAssembly. However, if one wanted to develop with e.g. Rust, which has tremendously good wasm support, we'd need other tools:

- install Rustc/Cargo via rustup
- add support for the `wasm32-wasi` target, with `rustup target add wasm32-wasi`, and can be tested by compiling a simple Hello world Rust program with `cargo build --target wasm32-wasi`.
- for client-side Web development based on Rust programs, `wasm-pack` might be useful: `cargo install wasm-pack`.

Tools that might be useful for any languages:

- I'd recommend installing `binaryen` too, which has a lot of useful tools for all sorts of things: https://github.com/WebAssembly/binaryen

Actually, because this image is quite targeted at C++ development, I'd recommend one of two strategies:

- have one Docker image per language for wasm development, which means more Dockerfiles to maintain, but then each can add their specific tooling without making the images crazy large. In this case, this specific Dockerfile we're looking at would be specialized for C++, and should be renamed/moved to a directory labelled as such.
- or have one Docker image for all of wasm development. This image might contain a lot of tools, especially if one thinks about adding support for e.g. .NET (Blazor) or other languages. But it's more convenient and the Docker image would be simpler to maintain.

Now this decision is up to you :) I'd go for the first approach myself, probably there's a way to share some common Docker image base across all the language-specific Docker images.

Hope this helps!

Reply:
@bnjbvr for wasm-pack see lines 5-6 also cargo-wasm also works for compiling to WASM

Reply:
I will look into binaryen thank-you so much for taking a look at this :heart: